src/components/Toc/Toc.tsx

```diff
@@ -25,50 +31,106 @@ export interface TocProps extends TocData {
     pdfLink?: string;
 }
 
-interface FlatTocItem {
-    name: string;
-    href?: string;
-    parents: string[];
-    opened?: boolean;
-}
-
 interface TocState {
-    flatToc: Record<string, FlatTocItem>;
-    filteredItemIds: string[];
-    filterName: string;
+    activeId: string | null | undefined;
+    fixedById: Record<string, 'opened' | 'closed'>;
     contentScrolled: boolean;
-    activeId?: string | null;
+}
+
+function linkTocItems(
+    parent: TocItem | null,
+    items: TocItem[],
+    itemById: Map<string, TocItem>,
+    parentById: Map<string, TocItem>,
+    itemIdByUrl: Map<string, string>,
+    singlePage?: boolean,
+) {
+    items.forEach((item) => {
+        itemById.set(item.id, item);
+
+        if (item.href) {
+            const {pathname, hash} = parse(item.href);
+            const url = singlePage ? normalizeHash(hash) : normalizePath(pathname);
+
+            itemIdByUrl.set(url as string, item.id);
+        }
+
+        if (parent) {
+            parentById.set(item.id, parent);
+        }
+
+        if (item.items) {
+            linkTocItems(item, item.items, itemById, parentById, itemIdByUrl, singlePage);
+        }
+    });
+}
+
+function getChildIds(item: TocItem): string[] {
+    return (item.items || ([] as TocItem[])).reduce((acc, child) => {
+        return acc.concat([child.id], getChildIds(child));
+    }, [] as string[]);
+}
+
+function getParentIds(item: TocItem, parents: Map<string, TocItem>) {
+    const result = [];
+
+    let parent = parents.get(item.id);
+    while (parent) {
+        result.push(parent.id);
+        parent = parents.get(parent.id);
+    }
+
+    return result;
 }
```

Comment:
May be all the helpers above should be moved to separate file to simplify current one?

And if I did it, it would be a class with all this internal maps and helpers.